### PR TITLE
Io1 2496/ Io1 2507: Constant log field names & debug field value functions not executed 

### PR DESCRIPTION
**Pull request by egorka-sereda**
https://bloxxx.atlassian.net/browse/IO1-2496
https://bloxxx.atlassian.net/browse/IO1-2507

* add log func for every heavy field. the field returned is a zap.Stringer, which will only evaluate if the log will be written
* change in code every log with heavy field to field func call
* add heavy fields logging rules to https://bloxxx.atlassian.net/browse/IO1-2496

### CODE REVIEW

**Comment by aaron-blox** [comment]
remove my commit

File: logging/lazy_fields.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+	"github.com/libp2p/go-libp2p/core/peer"
+
+	"github.com/bloxapp/ssv/network/records"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+var fieldMotLogged = "heavy field not logged"
```
Comment (aaron-blox) [inline logging/lazy_fields.go:28]:
```suggestion
var messageNotLogged = "heavy field not logged"
```

File: logging/lazy_fields.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+		return zap.String(key, strings.Join(v, ", "))
+	}
+
+	return zap.String(key, fieldMotLogged)
+}
+
+func showHeavyFields() bool {
+	if os.Getenv("SHOW_HEAVY_FIELDS") == "true" {
+		return true
+	}
```
Comment (aaron-blox) [inline logging/lazy_fields.go:183]: if we're being very careful about performance we shouldn't be looking up this value so much. it should probably a package level const/variable

**Review by aaron-blox — COMMENTED** [review]
in general looks good

**Review by aaron-blox — COMMENTED** [review]
please remove usages in all Info logs.

File: exporter/api/server.go [hunk 3]
```diff
@@ -89,7 +89,7 @@ func (ws *wsServer) BroadcastFeed() *event.Feed {
 func (ws *wsServer) RegisterHandler(endPoint string, handler func(conn *websocket.Conn)) {
 	ws.router.HandleFunc(endPoint, func(w http.ResponseWriter, r *http.Request) {
 		conn, err := upgrader.Upgrade(w, r, w.Header())
-		logger := ws.logger.With(zap.String("remote addr", conn.RemoteAddr().String()))
+		logger := ws.logger.With(zap.String("remote addr", conn.RemoteAddr().String())) // is it heavy field?
```
Comment (aaron-blox) [inline exporter/api/server.go:92]: please remove usages for logger.With
only usages should be with Debug lines

File: eth1/goeth/goETH.go [hunk 4]
```diff
@@ -228,7 +229,7 @@ func (ec *eth1Client) listenToSubscription(logs chan types.Log, sub ethereum.Sub
 				ec.logger.Warn("could not parse ongoing event, the event is malformed",
 					zap.String("event", eventName),
 					zap.Uint64("block", vLog.BlockNumber),
-					zap.String("txHash", vLog.TxHash.Hex()),
+					logging.LazyHash("txHash", vLog.TxHash),
```
Comment (aaron-blox) [inline eth1/goeth/goETH.go:232]: undo this change. it is for a Warn log

File: logging/LOGGING_RULES.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
+It's a draft of the logging rules for the project.
```
Comment (aaron-blox) [inline logging/LOGGING_RULES.md:1]: remove this file. add it as a comment/attachment to https://bloxxx.atlassian.net/browse/IO1-2512

File: logging/lazy_fields.go [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+func init() {
+	if os.Getenv("SHOW_HEAVY_FIELDS") == "true" {
+		showHeavyFields = true
+	}
+
+	//other checks, f.e. command line args or config file
+}
+
+func LazyBigInt(key string, val *big.Int) zapcore.Field {
```
Comment (aaron-blox) [inline logging/lazy_fields.go:41]: please write simple unit tests for all new functions

Comment (aaron-blox) [reply logging/lazy_fields.go:41]: for LazyBigInt the behavior was 
`zap.Uint64("syncOffset", syncOffset.Uint64()))`
but now you are using .String(). Is your implementation correct?

File: logging/lazy_fields.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+func LazyHash(key string, val common.Hash) zapcore.Field {
+	if showHeavyFields {
+		return zap.String(key, val.Hex())
+	}
+
+	return zap.String(key, messageNotLogged)
+}
+
+func LazyBytesHash(key string, val []byte) zapcore.Field {
```
Comment (aaron-blox) [inline logging/lazy_fields.go:57]: maybe rename something like LazyHexEncodeToString

BytesHash is very generic, what kind of hash is it?

File: logging/lazy_fields.go [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+	if showHeavyFields {
+		return zap.String(key, hex.EncodeToString(val))
+	}
+
+	return zap.String(key, messageNotLogged)
+}
+
+func LazyUrl(key string, val url.URL) zapcore.Field {
```
Comment (aaron-blox) [inline logging/lazy_fields.go:65]: I think many of these can be replaced by something like 
```suggestion
func LazyStringer(key string, str fmt.Stringer) zapcore.Field {
```

File: network/p2p/p2p.go [hunk 9]
```diff
@@ -188,7 +189,7 @@ func (n *p2pNetwork) startDiscovery() {
 			select {
 			case discoveredPeers <- e.AddrInfo:
 			default:
-				n.logger.Warn("connector queue is full, skipping new peer", zap.String("peerID", e.AddrInfo.ID.String()))
+				n.logger.Warn("connector queue is full, skipping new peer", logging.LazyPeerID("peerID", e.AddrInfo.ID))
```
Comment (aaron-blox) [inline network/p2p/p2p.go:192]: revert this change, it is a warn

File: network/peers/node_info_store.go [hunk 10]
```diff
@@ -30,7 +32,7 @@ func (pi *nodeInfoStore) Add(pid peer.ID, nodeInfo *records.NodeInfo) (bool, err
 		return false, errors.Wrap(err, "could not marshal node info record")
 	}
 	if err := pi.network.Peerstore().Put(pid, formatInfoKey(nodeInfoKey), raw); err != nil {
-		pi.logger.Warn("could not save peer data", zap.Error(err), zap.String("peer", pid.String()))
+		pi.logger.Warn("could not save peer data", zap.Error(err), logging.LazyPeerID("peer", pid))
```
Comment (aaron-blox) [inline network/peers/node_info_store.go:35]: revert this change it is a warn

File: operator/validator/controller.go [hunk 11]
```diff
@@ -664,7 +665,7 @@ func (c *controller) UpdateValidatorMetaDataLoop() {
 // SetupRunners initializes duty runners for the given validator
 func SetupRunners(ctx context.Context, logger *zap.Logger, options validator.Options) runner.DutyRunners {
 	if options.SSVShare == nil || options.SSVShare.BeaconMetadata == nil {
-		logger.Error("missing validator metadata", zap.String("validator", hex.EncodeToString(options.SSVShare.ValidatorPubKey)))
```
Comment (aaron-blox) [inline operator/validator/controller.go:667]: revert this change it is an Error

File: protocol/v2/ssv/validator/msgqueue_consumer.go [hunk 12]
```diff
@@ -32,7 +33,7 @@ func (v *Validator) HandleMessage(msg *spectypes.SSVMessage) {
 			v.logger.Warn("failed to decode message",
 				zap.Error(err),
 				zap.String("msg_type", message.MsgTypeToString(msg.MsgType)),
-				zap.String("msg_id", msg.MsgID.String()),
+				logging.LazyMessageID("msg_id", msg.MsgID),
```
Comment (aaron-blox) [inline protocol/v2/ssv/validator/msgqueue_consumer.go:36]: revert this change it is a warn

File: protocol/v2/ssv/validator/startup.go [hunk 13]
```diff
@@ -25,7 +26,7 @@ func (v *Validator) Start() error {
 			identifier := spectypes.NewMsgID(r.GetBaseRunner().Share.ValidatorPubKey, role)
 			if err := r.GetBaseRunner().QBFTController.LoadHighestInstance(identifier[:]); err != nil {
 				v.logger.Warn("failed to load highest instance",
-					zap.String("identifier", identifier.String()),
+					logging.LazyMessageID("identifier", identifier),
```
Comment (aaron-blox) [inline protocol/v2/ssv/validator/startup.go:29]: revert this change it is a warn

**Review by aaron-blox — COMMENTED** [review]
please make changes according to comments

**Review by aaron-blox — APPROVED** [review]

File: logging/lazy_fields.go [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+	"github.com/libp2p/go-libp2p/core/peer"
+
+	"github.com/bloxapp/ssv/network/records"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+const (
```
Comment (moshe-blox) [inline logging/lazy_fields.go:26]: Can we flip the names to start with `Field`?  I believe it's Go-ish & also good for autocomplete
```go
FieldAddress             = "address"
FieldBindIP              = "bindIP"
FieldBlockCacheMetrics   = "blockCacheMetricsField"
FieldCurrentSlot         = "currentSlot"
FieldDuration            = "duration"
FieldEnr                 = "enr"
FieldEventID             = "eventID"
FieldFromBlock           = "fromBlock"
FieldIdentifier          = "identifier"
FieldIndexCacheMetrics   = "indexCacheMetrics"
FieldMessageID           = "messageID"
FieldPeerID              = "peerID"
FieldPrivateKey          = "privKey"
FieldPubKey              = "pubKey"
FieldResults             = "results"
FieldStartTime           = "startTime"
FieldSubnets             = "subnets"
FieldSyncOffset          = "syncOffset"
FieldTargetNodeEnr       = "targetNodeEnr"
FieldTxHash              = "txHash"
FieldUpdatedEnrLocalNode = "updated_enr"
```

File: network/p2p/p2p_setup.go [hunk 15]
```diff
@@ -257,7 +257,7 @@ func (n *p2pNetwork) setupDiscovery() error {
 
 func (n *p2pNetwork) setupPubsub() error {
 	if n.cfg.PubSubTrace {
-		if err := logging.SetLogLevel("pubsub", zapcore.DebugLevel.String()); err != nil {
+		if err := golog.SetLogLevel("pubsub", zapcore.DebugLevel.String()); err != nil {
```
Comment (moshe-blox) [inline network/p2p/p2p_setup.go:260]: Let's reserve `golog` for this:
```
golog "log"
```

Maybe `ipfslog` or something?

**Review by moshe-blox — CHANGES_REQUESTED** [review]
Looks good, very clean API 🔥

**Comment by aaron-blox** [comment]
This design makes the code likely to run into circular dependencies. It caused one which I was able to easily fix by renaming a test package. 

Just something to be aware of as we continue refactoring the logging

**Review by aaron-blox — APPROVED** [review]

File: beacon/goclient/duties.go [hunk 16]
```diff
@@ -38,7 +40,7 @@ func (gc *goClient) GetDuties(epoch phase0.Epoch, validatorIndices []phase0.Vali
 	}
 	wg.Wait()
 
-	gc.logger.Debug("fetched duties", zap.Int("count", len(duties)), zap.Float64("duration (sec)", time.Since(start).Seconds()))
+	gc.logger.Debug("fetched duties", zap.Int("count", len(duties)), logging.DurationNano(start))
```
Comment (nivBlox) [inline beacon/goclient/duties.go:43]: should be DurationSecond

File: ibft/storage/store.go [hunk 17]
```diff
@@ -173,8 +173,7 @@ func (i *ibftStorage) CleanAllInstances(msgID []byte) error {
 	if err != nil {
 		return errors.Wrap(err, "failed to remove decided")
 	}
-	i.logger.Debug("removed decided", zap.Int("count", n),
-		zap.String("identifier", hex.EncodeToString(msgID)))
+	i.logger.Debug("removed decided", zap.Int("count", n), logging.IdentifierBytes(msgID))
```
Comment (nivBlox) [inline ibft/storage/store.go:176]: why not IdentifierString?

File: logging/fields.go [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+func BindIP(val net.IP) zapcore.Field {
+	return zap.Stringer(FieldBindIP, val)
+}
+
+func MessageID(val spectypes.MessageID) zapcore.Field {
+	return zap.Stringer(FieldMessageID, val)
+}
+
+func DurationNano(val time.Time) zapcore.Field {
+	return zap.Stringer(FieldDurationNano, int64Stringer{time.Since(val).Nanoseconds()})
```
Comment (nivBlox) [inline logging/fields.go:122]: why nano?

File: network/syncing/syncer.go [hunk 19]
```diff
@@ -207,9 +206,9 @@ func (s *syncer) getDecidedByRange(
 			})
 			logger.Debug("received and processed history batch",
 				zap.Int64("tail", int64(tail)),
-				zap.Duration("duration", time.Since(start)),
+				logging.DurationNano(start),
 				zap.Int("results_count", len(msgs)),
-				zap.String("results", s.formatSyncResults(msgs)),
+				logging.Results(msgs),
```
Comment (nivBlox) [inline network/syncing/syncer.go:211]: too vague. need to be based on what the result type is

File: protocol/v2/message/consensus_test.go [hunk 20]
```diff
@@ -1,9 +1,10 @@
-package message
+package message_test
```
Comment (nivBlox) [inline protocol/v2/message/consensus_test.go:1]: why changed package?   should be in the same package with consensus.go

**Review by nivBlox — CHANGES_REQUESTED** [review]

File: exporter/api/test_utils.go [hunk 21]
```diff
@@ -66,7 +67,7 @@ func (client *WSClient) StartStream(addr, path string) error {
 // StartQuery initiates query requests
 func (client *WSClient) StartQuery(addr, path string) error {
 	u := url.URL{Scheme: "ws", Host: addr, Path: path}
-	client.logger.Debug("connecting to server", zap.String("addr", u.String()))
+	client.logger.Debug("connecting to server", logging.LazyUrl("addr", u))
```
[the remark this answers is not in the document]
Comment (aaron-blox) [reply exporter/api/test_utils.go]: has been changed to `client.logger.Debug("connecting to server", logging.Address(u))` to capture the field name

File: ibft/storage/store.go (hunk 17 above)
Comment (aaron-blox) [reply ibft/storage/store.go:176]: the naming convention i used was based on the data type of the function argument. Because the messageID here is a byte slice, the name was made IdentifierBytes. 

if we have a naming convention that is output based, like we name it based on the output type of the function, we will run into the issue where names collide 

```go

func IdentifierString(id []bytes) zap.Field {
  return a string of hex encoded bytes 
} 


func IdentifierString(id string) zap.Field {
  return the id which is already a string
} 

```

File: protocol/v2/message/consensus_test.go (hunk 20 above)
Comment (aaron-blox) [reply protocol/v2/message/consensus_test.go:1]: these changes caused one import cycle and are likely to cause more, later we're gonna document about that in some detail here https://blox-ensemble.monday.com/docs/4034094011

what happens is the logging package with the functions that return zap.Fields import type definitions from all over the project. the specific import cycle this caused could be easily fixed by using the `_test` suffix package name for the test files.
 
In this example there is a circular dependency via:
`logging` imports `types`
`types` imports `ssv`
`ssv` imports `logging`

```go
package logging

import (
  "ssv/types"
)

func IDString(id types.ID) zap.Field {
  return ...
}
```

```go
package types

import (
  "ssv/node"
)

type Something struct {
  node node.Node
}
```


```go
package node

import (
  "ssv/logging"
  "ssv/types"
)

func doWork() {
  logger.Info("msg", logging.IDString(someID))
}

File: logging/fields.go (hunk 18 above)
Comment (aaron-blox) [reply logging/fields.go:122]: it's the smallest available value. for example if you used milliseconds and are debugging a timing issue in prod, it can appear that something is happening at the same time because the actions happened in the same millisecond, but in reality one action preceded the other by a duration smaller than one millisecond

File: beacon/goclient/duties.go (hunk 16 above)
Comment (aaron-blox) [reply beacon/goclient/duties.go:43]: is using seconds prefered for any reason?

File: network/syncing/syncer.go (hunk 19 above)
Comment (aaron-blox) [reply network/syncing/syncer.go:211]: would something like "sync_results" be good?

File: beacon/goclient/duties.go (hunk 16 above)
Comment (nivBlox) [reply beacon/goclient/duties.go:43]: I think nano is too small scope. MiliSec is better IMO

File: network/syncing/syncer.go (hunk 19 above)
Comment (nivBlox) [reply network/syncing/syncer.go:211]: yeah, better

**Comment by aaron-blox** [comment]
All comments resolved, i think this is okay to merge
